TwoNote/app/src/main/java/com/microsoft/device/display/samples/twonote/includes/FileHandler.kt

```diff
@@ -0,0 +1,119 @@
+    // defaults to public Downloads folder if uri is empty
+    fun createFile(pickerInitialUri: Uri) {
+        val intent = Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
+            addCategory(Intent.CATEGORY_OPENABLE)
+            type = Defines.PLAIN_TEXT
+            putExtra(Intent.EXTRA_TITLE, Defines.DEFAULT_FILE_NAME)
+
+            putExtra(DocumentsContract.EXTRA_INITIAL_URI, pickerInitialUri)
+        }
+        startActivityForResult(activity, intent, CREATE_FILE, null)
+    }
+
+    // creating a window prompting user to choose a file to open
+    // defaults to public Downloads folder if uri is empty
+    fun openFile(pickerInitialUri: Uri) {
+        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
+            addCategory(Intent.CATEGORY_OPENABLE)
+            type = Defines.PLAIN_TEXT
+
+            putExtra(DocumentsContract.EXTRA_INITIAL_URI, pickerInitialUri)
+        }
+
+        startActivityForResult(activity, intent, PICK_TXT_FILE, null)
+    }
+
+    // read text from file specified in uri path
+    @Throws(IOException::class)
+    private fun readTextFromUri(uri: Uri, event: DragEvent?): String {
+        val stringBuilder = StringBuilder()
+        ActivityCompat.requestDragAndDropPermissions(activity, event)
+        contentResolver.openInputStream(uri)?.use { inputStream ->
+            BufferedReader(InputStreamReader(inputStream)).use { reader ->
+                var line: String? = reader.readLine()
+                while (line != null) {
+                    stringBuilder.append(line)
+                    line = reader.readLine()
+                }
+            }
+        }
+        return stringBuilder.toString()
+    }
+
+    // overwrite text from file specified in uri path
+    fun alterDocument(uri: Uri) {
+        try {
+            contentResolver.openFileDescriptor(uri, "w")?.use {
+                FileOutputStream(it.fileDescriptor).use { stream ->
+                    val charset: Charset = Charsets.UTF_8
+                    stream.write(
+                        textField.getText().toString()
+                            .toByteArray(charset)
+                    )
+                }
+            }
+        } catch (e: FileNotFoundException) {
+            e.printStackTrace()
+        } catch (e: IOException) {
+            e.printStackTrace()
+        }
+    }
+
+    // format text for readability (newline chars are dropped in saving/grabbing process)
+    fun processFileData(uri: Uri, event: DragEvent?) {
+        val str: String = readTextFromUri(uri, event)
+
+        val builder = StringBuilder()
+        var initHeader = true
+
+        val lines = str.split("<")
+        lines.forEach {
+            if (initHeader) {
+                builder.append(it)
+                initHeader = false
+            } else {
+                builder.append("<" + it + System.getProperty("line.separator"))
+            }
+        }
+        textField.setText(builder.toString())
```

Comment:
maybe append the text instead of replacing it all?

Reply:
I don't know if it's necessary to do now since we'll have to change the way that we handle drag and drop anyways.